### PR TITLE
[SPARK-43393][SQL] Address sequence expression overflow bug.

### PR DESCRIPTION
**Pull request by thepinetree**
### What changes were proposed in this pull request?
Spark has a (long-standing) overflow bug in the `sequence` expression.

Consider the following operations:
```
spark.sql("CREATE TABLE foo (l LONG);")
spark.sql(s"INSERT INTO foo VALUES (${Long.MaxValue});")
spark.sql("SELECT sequence(0, l) FROM foo;").collect()
```

The result of these operations will be:
```
Array[org.apache.spark.sql.Row] = Array([WrappedArray()])
```
an unintended consequence of overflow.

The sequence is applied to values `0` and `Long.MaxValue` with a step size of `1` which uses a length computation defined [here](https://github.com/apache/spark/blob/16411188c7ba6cb19c46a2bd512b2485a4c03e2c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala#L3451). In this calculation, with `start = 0`, `stop = Long.MaxValue`, and `step = 1`, the calculated `len` overflows to `Long.MinValue`. The computation, in binary looks like:

```
  0111111111111111111111111111111111111111111111111111111111111111
- 0000000000000000000000000000000000000000000000000000000000000000 
------------------------------------------------------------------
  0111111111111111111111111111111111111111111111111111111111111111
/ 0000000000000000000000000000000000000000000000000000000000000001
------------------------------------------------------------------
  0111111111111111111111111111111111111111111111111111111111111111
+ 0000000000000000000000000000000000000000000000000000000000000001
------------------------------------------------------------------
  1000000000000000000000000000000000000000000000000000000000000000
```

The following [check](https://github.com/apache/spark/blob/16411188c7ba6cb19c46a2bd512b2485a4c03e2c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala#L3454) passes as the negative `Long.MinValue` is still `<= MAX_ROUNDED_ARRAY_LENGTH`. The following cast to `toInt` uses this representation and [truncates the upper bits](https://github.com/apache/spark/blob/16411188c7ba6cb19c46a2bd512b2485a4c03e2c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala#L3457) resulting in an empty length of `0`.

Other overflows are similarly problematic.

This PR addresses the issue by checking numeric operations in the length computation for overflow.

### Why are the changes needed?
There is a correctness bug from overflow in the `sequence` expression.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Tests added in `CollectionExpressionsSuite.scala`.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @gengliangwang FYI

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3470]: let's use explicit `if` to throw the error.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3474]: if it's not reachable, let's throw `SparkException.internalError`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3470]: let's use explicit `if` to throw the error.

**Review by cloud-fan — APPROVED** [review]
LGTM if tests pass

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3462]: If the exception is an user-facing error, let's introduce an error class, and raise a Spark exception w/ it.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala
Comment (thepinetree) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Ah, I originally misunderstood your comment and with @ankurdave's help I learned about the errors defined in `QueryExecutionErrors.scala` and that they are actually used in this file. I thought `createArrayWithElementsExceedLimitError` might be the most fitting.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3452]: is it possible to add a util function for the code below? then codegen can just call it. we can put the util function in `object Sequence`

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3452]: The util function can just take long type inputs, and caller side is responsible to turn `start`/`stop`/`estimatedStep` to long.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3146]: maybe just use an assert? Assertion error is also treated as internal errors.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3534]: I'm surprised that this works. Should we do `classOf[Sequence].getName` + ".sequenceLength"?

Comment (thepinetree) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3534]: Not sure exactly how, but my hypothesis is that these factors all play a role:
* Sequence object ends up in the same compilation unit as this function (not sure if this is expected)
* `sequenceLength` function is effectively static in Scala and publicly accessible

I noticed some other functions in this file do this as well -- e.g. https://github.com/apache/spark/blob/128f5523194d5241c7b0f08b5be183288128ba16/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala#L613

I do like this better though, cleaner and easier to understand.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala
Comment (thepinetree) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:3146]: I personally like the current exception better since it's more descriptive of the actual problem -- trying to create too large an array (with the user's intended size) and what the limit is. If strong opinion, I can change to an assertion.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.5!

**Comment by dongjoon-hyun** [comment]
Thank you, @thepinetree and @cloud-fan . Given that this is a long-standing overflow bug, do you think we can have this fix in other live release branches, `branch-3.4` and `branch-3.3`? Especially, I'm interested in `branch-3.4` as the release manager of Apache Spark 3.4.2.

**Comment by dongjoon-hyun** [comment]
Oh this seems to break branch-3.5.
- https://github.com/apache/spark/actions/runs/6873765275

Let me revert this from branch-3.5. Given the situation, we can start backport from branch-3.5 to branch-3.3 as three separate PRs, @thepinetree .

**Comment by thepinetree** [comment]
@dongjoon-hyun @cloud-fan 

Backport PRs:
* 3.3: https://github.com/apache/spark/pull/43821
* 3.4: https://github.com/apache/spark/pull/43819
* 3.5: https://github.com/apache/spark/pull/43820

**Comment by dongjoon-hyun** [comment]
Thank you so much!

**Comment by dongjoon-hyun** [comment]
Could you fix the compilation of your PRs, @thepinetree ?

**Comment by dongjoon-hyun** [comment]
Gentle ping, @thepinetree . All backporting PRs are broken at the compilation stage.

**Comment by thepinetree** [comment]
Hi @dongjoon-hyun! Yes, sorry, I thought they'd be a simple back port apart from a simple import conflict. I'll fix the compilation errors appropriately when I have a chance tonight.

**Comment by dongjoon-hyun** [comment]
Thank you so much, @thepinetree !

**Review by beliefer — COMMENTED** [review]
late LGTM.

**Comment by thepinetree** [comment]
quick update @dongjoon-hyun, looks like the 3.4/3.5 backports should be good to go after some flakiness is resolved (documentation and one spark connect suite).

3.4/3.5 had an older version of the errors (same across both versions) and I made changes accordingly.
3.3 had even older error definitions and needed some more changes on top of that.

**Pull request closed without merging**
